mp4/aac.go

```diff
@@ -0,0 +1,164 @@
+package mp4
+
+import (
+	"errors"
+	"io"
+
+	"github.com/edgeware/mp4ff/bits"
+)
+
+const (
+	// AAClc - AAC-LC Low Complexity
+	AAClc = 2
+	// HEAACv1 - HE-AAC version 1 with SBR
+	HEAACv1 = 5
+	// HEAACv2 - HE-AAC version 2 with SBR and PS
+	HEAACv2 = 29
+)
+
+// AudioSpecificConfig according to ISO/IES 14496-3
+// Syntax specified in Table 1.15
+type AudioSpecificConfig struct {
+	ObjectType           byte
+	ChannelConfiguration byte // Defined in Table 1.19
+	SamplingFrequency    int
+	ExtensionFrequency   int
+	SBRPresentFlag       bool
+	PSPresentFlag        bool
+}
+
+var frequencyTable = map[byte]int{
+	0:  96000,
+	1:  88200,
+	2:  64000,
+	3:  48000,
+	4:  44100,
+	5:  32000,
+	6:  24000,
+	7:  22050,
+	8:  16000,
+	9:  12000,
+	10: 11025,
+	11: 8000,
+	12: 7350,
+}
+
+var reverseFrequencies = map[int]byte{
+	96000: 0,
+	88200: 1,
+	64000: 2,
+	48000: 3,
+	44100: 4,
+	32000: 5,
+	24000: 6,
+	22050: 7,
+	16000: 8,
+	12000: 9,
+	11025: 10,
+	8000:  11,
+	7350:  12,
+}
+
+// DecodeAudioSpecificConfig -
+func DecodeAudioSpecificConfig(r io.Reader) (AudioSpecificConfig, error) {
+	rb := bits.NewReader(r)
+
+	asf := AudioSpecificConfig{}
+	audioObjectType := byte(rb.MustRead(5))
+	switch audioObjectType {
+	case AAClc, HEAACv1, HEAACv2:
+		// All fine
+	default:
+		return asf, errors.New("Only LC, HE-AACv1, and HE-AACv2 supported")
+	}
+	asf.ObjectType = audioObjectType
+	frequency, ok := getFrequency(rb)
+	if !ok {
+		return asf, errors.New("Strange frequency index")
+	}
+	asf.SamplingFrequency = frequency
+	asf.ChannelConfiguration = byte(rb.MustRead(4))
+	switch audioObjectType {
+	case HEAACv1, HEAACv2:
+		frequency, ok := getFrequency(rb)
+		if !ok {
+			return asf, errors.New("Strange frequency index")
+		}
+		asf.ExtensionFrequency = frequency
+		audioObjectType = byte(rb.MustRead(4))
+		if audioObjectType == 22 {
+			return asf, errors.New("ExtensionChannelConfiguration not supported")
+		}
+	}
+	switch audioObjectType {
+	case AAClc:
+		// Read GASpecificConfig - Table 4.1 in ISO/IEC 14496-3 part 4
+		frameLengthFlag := rb.MustReadFlag()
+		if !frameLengthFlag {
+			return asf, errors.New("Does not support frameLengthFlag = 1")
+		}
+		dependsOnCoreCoder := rb.MustReadFlag()
+		if dependsOnCoreCoder {
+			return asf, errors.New("Does not support dependsOnCoreCoder")
+		}
+		extensionFlag := rb.MustReadFlag()
+		if extensionFlag {
+			return asf, errors.New("Does not support dependsOnCoreCoder")
+		}
+	default:
+		panic("Cannot handle audioObjectType")
+	}
+	// Done (there may be trailing bits)
```

Comment:
Won't trailing bits cause issues with the next read?

Reply:
No, because each mp4 box has a length so when we read the next box we start at its offset.

It is very complex to parse the full AAC bitstream, so we just parse down and get the info we want and can handle.